Implement Bech32 encoding/decoding

linked0 opened the pull request:
These are utilities for converting public keys of `ubyte` array type and bech32-encoded strings. We should switch the address format from Stellar-style to bech32. So this is the first step for the task.

This PR needs more comments and descriptions. And it also needs more tests. So it is now a WIP state.

Fixes #1655

AndrejMitrovic commented on source/agora/common/Bech32.d:
```diff
@@ -0,0 +1,232 @@
+ *  bits correspond to earlier values. */
+uint polymod(const ubyte[] values)
+{
+    // The input is interpreted as a list of coefficients of a polynomial over F = GF(32), with an
```
This was clearly taken from some other code. Please keep existing copyrights in place.

linked0 replied:
I will. Thanks.

linked0 committed:
Implement Bech32 encoding/decoding

This implementation was ported from the code written in C++ to D language. The
original code can be found at https://github.com/sipa/bech32/tree/master/ref/c++

linked0 committed:
Apply Bech32 encoding to `PublicKey`

We use Bech32 instead of Base32 in encoding/decoding `PublicKey`.

linked0 commented:
This PR now only passes in unit tests of `agora.crypto.Key`. In order to make all the unit tests passed, we should change all the key pairs of the `WellKnownKeys` module. So, I'd like to have a whiteboard or pair-programming about it. And there are something more to share and discuss.
@Geod24 I will ask you for a whiteboard tomorrow. Thanks.

linked0 commented:
Need to rebase and some updates

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+import agora.crypto.ECC;
+
+/// The Bech32 character set for encoding.
+const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
```
```suggestion
private immutable string CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+const char* CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
+
+/// The Bech32 character set for decoding.
+const byte[128] CHARSET_REV = [
```
```suggestion
private immutable byte[128] CHARSET_REV = [
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+
+*******************************************************************************/
+
+uint polymod (const ubyte[] values)
```
```suggestion
private uint polymod (const ubyte[] values) @safe pure nothrow @nogc
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+
+*******************************************************************************/
+
+ubyte[] expand_hrp (const string hrp)
```
```suggestion
private ubyte[] expand_hrp (in char[] hrp) @safe pure nothrow
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+        ret[i] = c >> 5;
+        ret[i + hrp.length + 1] = c & 0x1f;
+    }
+    ret[hrp.length] = 0;
+    return ret;
```
So are we returning a string here ? The 0 termination would imply so.
Also we should probably return a slice, not just `ret`. This code seems buggy to me.

linked0 replied:
I see what you mean. Will update.

linked0 replied:
By analyzing it, it turns out that the 0 does not mean termination because it's not `char` array, but `ubyte` array.

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+/*******************************************************************************
```
Could you move the public functions at the top of this module ?
I expect there's one to encode, one to decode, and they should be the first thing one see when opening this module.

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+/*******************************************************************************
+
+    Utilities for converting between bech32-encoded string and `ubyte` array.
+
```
```suggestion

    See_Also:
        https://github.com/sipa/bech32
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+
+*******************************************************************************/
+
+ubyte[] create_checksum (string hrp, const ubyte[] values)
```
```suggestion
ubyte[] create_checksum (in char[] hrp, in ubyte[] values)
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+ubyte[] create_checksum (string hrp, const ubyte[] values)
+{
+    ubyte[] enc = expand_hrp(hrp);
+    enc ~= cast(ubyte [])values;
```
The cast shouldn't be needed ?

linked0 replied:
Right. It doesn't need it.

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+    ubyte[] enc = expand_hrp(hrp);
+    enc ~= cast(ubyte [])values;
+    enc.length = enc.length + 6;
+    uint mod = polymod(enc) ^ 1;
```
Did you find a good resource about Bech32M ?

Geod24 replied:
https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki

linked0 replied:
Thanks for the information.

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,486 @@
+{
+    ubyte[] expected = [3, 3, 3, 0, 2, 15, 1];
+    auto expanded = expand_hrp("boa");
+    assert(expanded == expected);
+}
+
+unittest
+{
+    import std.stdio;
+
+    static struct ValidAddressData
+    {
+        public string address;
+        public Point pubkey;
+    }
+
+    ValidAddressData[] valid_addresses = [
+        // CoinNet Genesis Address
+        // GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ
+        ValidAddressData(
+            "boa1xzwsyw8q59c5qz7x669fmxe3dtx4zztyjyf6qhpgfapfd54nqy302ft5v67",
+            Point(cast(ubyte[])[
+                0x9D, 0x02, 0x38, 0xE0, 0xA1, 0x71, 0x40, 0x0B,
+                0xC6, 0xD6, 0x8A, 0x9D, 0x9B, 0x31, 0x6A, 0xCD,
+                0x51, 0x09, 0x64, 0x91, 0x13, 0xA0, 0x5C, 0x28,
+                0x4F, 0x42, 0x96, 0xD2, 0xB3, 0x01, 0x22, 0xF5])
+        ),
```
Those are self-referential. Would be better to use the reference implementation tests here: https://github.com/sipa/bech32/blob/master/ref/c%2B%2B/tests.cpp

linked0 replied:
Can you explain more about what you mean? I will look into it more anyway.

Geod24 replied:
Those are not good tests, because they are based on the implementation, rather than written independently.
Writing an implementation then adding tests that match it doesn't confirm that we conform to the standard.
Instead, the D implementation should be checked against the same tests that the C++ implementation is using.

linked0 replied:
Added more test cases.

linked0 commented:
Ready for review.

Geod24 commented on source/scpd/types/Stellar_SCP.d:
```diff
@@ -321,41 +321,42 @@ struct SCPQuorumSet {
 
     assert(qc1.hashFull() == Hash.fromString(
-        "0x7b56982f02dfbf3737ff560fe6511674e182d38475f0516cb8b0f338a3156543b0731e66c9a3ced01c238652b12e51c95207ec2bf6eae237f24b08a357a1bd2a"),
+        "0x2a37ad395b2c50d18f2a52e4b6759a8bc2e43f7835ef5f41805d1f717e528943f0de5a15ec9bf4916db6da76f4d1f04e7df0bfdd5d5973a734acce115f07ed67"),
```
Why did the hash change ? The quorum set hash only depends on the binary representation of the public key, doesn't it ?

linked0 replied:
Will update.

Geod24 commented on source/scpd/types/Stellar_SCP.d:
```diff
@@ -321,41 +321,42 @@ struct SCPQuorumSet {
-        "0x7b56982f02dfbf3737ff560fe6511674e182d38475f0516cb8b0f338a3156543b0731e66c9a3ced01c238652b12e51c95207ec2bf6eae237f24b08a357a1bd2a"),
+        "0x2a37ad395b2c50d18f2a52e4b6759a8bc2e43f7835ef5f41805d1f717e528943f0de5a15ec9bf4916db6da76f4d1f04e7df0bfdd5d5973a734acce115f07ed67"),
         qc1.hashFull().to!string);
+    import std.stdio;
```
Leftover

Geod24 commented on source/agora/utils/SCPPrettyPrinter.d:
```diff
@@ -448,8 +448,8 @@ unittest
     auto qc = QuorumConfig(2,
-        [PublicKey.fromString("GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN"),
-         PublicKey.fromString("GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5")]);
+        [PublicKey.fromString("boa1xrv266cegdthdc87uche9zvj8842shz3sdyvw0qecpgeykyv4ynssf79ndd"),
+         PublicKey.fromString("boa1xzq266cegdthdc87uche9zvj8842shz3sdyvw0qecpgeykyv4ynss3dne57")]);
```
Those two values seem awfully close to one another, how did you obtain them ?

linked0 replied:
Only the first byte is different for each public address. The first address has the value, 216 for the first byte, and the second one has 128 for that.

Geod24 replied:
Please use the previous key's value so we don't have to update the hash, just the string representation.

linked0 replied:
Ah, OK. I will.

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,757 @@
+    Invalid,
+    Bech32,  // Bech32 encoding as defined in BIP173
+    Bech32m, // Bech32m encoding as defined in BIP350
+};
```
```suggestion
}
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,757 @@
+    Encoding encoding;  // What encoding was detected in the result
+    char[] hrp;         // The human readable part
+    ubyte[] data;       // The payload (excluding checksum)
+};
```
```suggestion
}
```

Geod24 commented on source/agora/crypto/Key.d:
```diff
@@ -134,27 +137,30 @@ public struct PublicKey
-        bin[$ - ChecksumWidth .. $] = checksum(bin[0 .. $ - ChecksumWidth]);
-        Base32.encode(bin, sink);
+        bin[VersionWidth .. $] = this.data[];
+        string encoded = encodeBech32(HumanReadablePart, bin,
+            Encoding.Bech32m, true).assumeUnique;
+        foreach (enc_char; encoded)
+            sink([enc_char]);
```
The point of this overload is to avoid allocation, and this code completely fails this.
First and foremost, why the foreach ? It's iterating over *every element of the array* then allocating a *1-element array* to pass to the sink. Why not feed `encoded` directly to the sink ???

`encodeBech32` could be made to not allocate, but we will leave this for another PR. But I really can't understand what's the point of this `foreach`...

linked0 replied:
This is because I'm not familiar with the concept of the `sink`. I will look into the use of the `sink` and modify it right now.

Geod24 commented on source/agora/crypto/Key.d:
```diff
@@ -134,27 +137,30 @@ public struct PublicKey
     public string toString () const @trusted nothrow
     {
-        ubyte[VersionWidth + PublicKey.sizeof + ChecksumWidth] bin;
+        scope(failure) assert(0);
```
Why ?

linked0 replied:
The `encodeBech32` was not `nothrow` in the previous code, so this is leftover. I removed it.

Geod24 commented on source/agora/crypto/Key.d:
```diff
@@ -112,6 +113,8 @@ unittest
 
 private immutable int VersionWidth = 1;
 private immutable int ChecksumWidth = 2;
+private immutable int HumanReadableWidth = 3;
```
Remove and use `.length`

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,757 @@
+
+    ];
+
+    foreach(input; addresses_bech32_m)
```
```suggestion
    foreach (input; addresses_bech32_m)
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,757 @@
+
+    ];
+
+    foreach(input; addresses_bech32)
```
```suggestion
    foreach (input; addresses_bech32)
```

Geod24 commented on source/agora/crypto/Bech32.d:
```diff
@@ -0,0 +1,757 @@
+    return true;
+}
+
+// Convert a input string to a lower case string
+version(unittest) char[] toLowerString(const(char)[] input) pure nothrow @safe
+{
+    char[] conv;
+    foreach (c; input)
+        conv ~= toLower(c);
+    return conv;
+}
```
There's a function in the std library for this

linked0 replied:
> There's a function in the std library for this

You mean this function, [toLower](https://dlang.org/library/std/uni/to_lower.html#1)?

Geod24 replied:
Yes

linked0 replied:
Updated, Ty.

linked0 commented:
Ready for review

Geod24 commented on source/agora/utils/PrettyPrinter.d:
```diff
@@ -172,22 +172,22 @@ private struct PubKeyFmt
-            enum EndFrom    = 56 - 4;
-            size_t count;
+            // Public keys are 63 characters, only take the first 12 and last 4
+            // Only format `0xboa1acdefghi..6789`
```
```suggestion
            // Only format `boa1acdefghi..6789`
```

Geod24 replied:
Actually I wonder if we should just not print the HRP ?

linked0 replied:
I thought about that also. Can't we use it as identifying the network like bitcoin's use case, like `bt`, `tb`?  Anyway, we can remove that now and restore it later if needed.

Geod24 reviewed (approved):
LGTM, just fix the one comment and we can merge.

AndrejMitrovic commented:
Does this mean we have to mine the keys again?
